### PR TITLE
Closes #352: Expose tracking protection in session/engine

### LINKED ISSUE #352: Expose tracking protection in session/engine
The goal of this issue is working towards getting browser-session into a state where we can use it in Firefox for Fire TV (see [Google doc](https://docs.google.com/spreadsheets/d/17a4bIYB88eovF_UwiaXjL3uJa3r7Fidl8IzAzV6QEbw/edit#gid=0)). Tracking protection is the last big missing piece.

In GeckoView enabling tracking protection is done automatically for private browsing. In the WebView version of Focus we do this manually by intercepting URLs from the TP lists.

In Focus' version of `session` we save whether tracking protection is enabled and how many trackers have been blocked for the currently loaded page (That's something we do not need in Fire TV; only later for Focus).

### PR DESCRIPTION
**Pull request by csadilek**
This exposes tracking protection in our session/engine. It's currently only implemented for GeckoEngine. I suggest filing a follow-up for making this work with WebView so we can make some progress?

### CODE REVIEW

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 1]
```diff
@@ -40,6 +42,8 @@ class Session(
         fun onSecurityChanged(session: Session, securityInfo: SecurityInfo) = Unit
         fun onCustomTabConfigChanged(session: Session, customTabConfig: CustomTabConfig?) = Unit
         fun onDownload(session: Session, download: Download): Boolean = false
+        fun onTrackerBlockingEnabledChanged(session: Session, blockingEnabled: Boolean): Unit = Unit
```
Comment (sputnikci) [inline components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: \[ktlint\] WARNING: \[no\-unit\-return\] Unnecessary "Unit" return type in column 90

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 2]
```diff
@@ -40,6 +42,8 @@ class Session(
...
+        fun onTrackerBlocked(session: Session, blocked: String, all: List<String>): Unit = Unit
```
Comment (sputnikci) [inline components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: \[ktlint\] WARNING: \[no\-unit\-return\] Unnecessary "Unit" return type in column 85

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 3]
```diff
@@ -11,23 +11,30 @@ import org.mozilla.geckoview.GeckoResponse
...
+import org.mozilla.geckoview.GeckoSessionSettings
 
 /**
  * Gecko-based EngineSession implementation.
  */
+@Suppress("TooManyFunctions")
 class GeckoEngineSession(
-    runtime: GeckoRuntime
+    private val runtime: GeckoRuntime,
+    privateMode: Boolean = false
```
Comment (pocmo) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: I think we should keep private browsing and tracking protection separated. Is this possible with GeckoView?

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 4]
```diff
@@ -19,17 +19,17 @@ class GeckoEngine(
...
     }
 
     /**
-     * Create a new Gecko-based EngineSession.
+     * Creates a new Gecko-based EngineSession.
      */
-    override fun createSession(): EngineSession {
-        return GeckoEngineSession(runtime)
+    override fun createSession(private: Boolean): EngineSession {
+        return GeckoEngineSession(runtime, private)
```
Comment (pocmo) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: Eventually we may want to introduce some kind of `EngineSessionSettings`?

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 5]
```diff
@@ -174,6 +178,20 @@ class Session(
...
+        notifyObservers(old, new) { onTrackerBlockingEnabledChanged(this@Session, trackerBlockingEnabled) }
+    }
+
+    /**
+     * List of URIs that have been blocked in this session.
+     */
+    var trackersBlocked: List<String> by Delegates.observable(emptyList()) { _, old, new ->
+        notifyObservers(old, new) { onTrackerBlocked(this@Session, trackersBlocked.last(), trackersBlocked) }
+    }
+
```
Comment (pocmo) [inline components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: So far we only need the `count` in our apps and not the URLs.

File: components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt [hunk 6]
```diff
@@ -39,6 +40,14 @@ internal class EngineObserver(val session: Session) : EngineSession.Observer {
                 ?: "", issuer ?: "")
     }
 
+    override fun onTrackerBlocked(url: String) {
+        session.trackersBlocked += url
+    }
+
```
Comment (pocmo) [inline components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt]: This information is dependent on the page loaded and we clear that in Focus whenever we load a new page.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 3 above)
Comment (csadilek) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: OK, let's postpone private browsing. I will remove the commit.

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt (hunk 5 above)
Comment (csadilek) [reply components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: I saw, yes, but having the list seems like a better solution. It doesn't require additional logic and we can always get the count from the list. It also seemed likely to me that we want to be able to tell *what* was blocked. So, my preference would be to keep the list.

File: components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt (hunk 6 above)
Comment (csadilek) [reply components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt]: OK, I will update and add this.

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt (hunk 5 above)
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: Alright. We may be interested in the blocked _categories_ soon. But the GV API doesn't give us that for now anyways.

**Pull request merged**